src/proc/syncer-processing-block.cpp

```diff
@@ -10,63 +10,98 @@
+        bool enabled = false;
+        size_t n_opts = 0;
+        for( auto & wopt : _enable_opts )
         {
-            std::stringstream ss;
-            ss << "SYNCED: ";
-            auto composite = dynamic_cast<composite_frame*>(f.frame);
-            for (int i = 0; i < composite->get_embedded_frames_count(); i++)
+            auto opt = wopt.lock();
+            if( opt )
             {
-                auto matched = composite->get_frame(i);
-                ss << matched->get_stream()->get_stream_type() << " " << matched->get_frame_number() << ", "<<std::fixed<< matched->get_frame_timestamp()<<" ";
+                ++n_opts;
+                if( opt->is_true() )
+                {
+                    enabled = true;
+                    break;
+                }
             }
+        }
+        if( n_opts && ! enabled )
+        {
+            get_source().frame_ready( std::move( frame ) );
+            return;
+        }
 
-            LOG_DEBUG(ss.str());
-            env.matches.enqueue(std::move(f));
-        });
-
-        auto f = [&](frame_holder frame, synthetic_source_interface* source)
         {
-            // if the syncer is disabled passthrough the frame
-            bool enabled = false;
-            size_t n_opts = 0;
-            for( auto& wopt : _enable_opts )
+            std::lock_guard< std::mutex > lock( _mutex );
+
+            if( _matcher == nullptr )
             {
-                auto opt = wopt.lock();
-                if( opt )
-                {
-                    ++n_opts;
-                    if( opt->is_true() )
-                    {
-                        enabled = true;
-                        break;
-                    }
-                }
+                create_matcher( frame, log );
             }
-            if( n_opts  &&  ! enabled )
+
+            _matcher->dispatch( std::move( frame ), { source, matches, log } );
+        }
+
+        frame_holder f;
+        {
+            std::lock_guard< std::mutex > lock(callback_mutex);
+
+            while( matches.try_dequeue( &f ) )
             {
-                get_source().frame_ready( std::move( frame ) );
-                return;
+                get_source().frame_ready( std::move( f ) );
             }
+        }
+    };
 
-            single_consumer_frame_queue<frame_holder> matches;
+    set_processing_callback( std::shared_ptr< rs2_frame_processor_callback >(
+        new internal_frame_processor_callback< decltype( f ) >( f ) ) );
+}
 
+void syncer_process_unit::create_matcher( const frame_holder & frame, bool log )
+{
+    auto sensor = frame.frame->get_sensor().get();
+    const device_interface * dev = nullptr;
+    try
+    {
+        dev = sensor->get_device().shared_from_this().get();
+    }
+    catch( const std::bad_weak_ptr & )
+    {
+        LOG_WARNING( "Device destroyed" );
+    }
+    if( dev )
+    {
+        _matcher = dev->create_matcher( frame );
```

Comment:
Move to `try` block?

Reply:
Consider removing the `dev->create_matcher...` code from sync.cpp

Reply:
done